mypy/stubgenc.py

```diff
@@ -189,6 +179,86 @@ def generate_c_function_stub(module: ModuleType,
             ))
 
 
+def _infer_signature_for_c_function_stub(
+        class_name: Optional[str],
+        class_sigs: Dict[str, str],
+        name: str,
+        obj: object,
+        sigs: Dict[str, str]) -> List[FunctionSig]:
+    default_ret_type = 'None' if name == '__init__' and class_name else 'Any'
+
+    if type(obj).__name__ == 'cython_function_or_method':
+        # Special-case Cython functions: if binding=True when compiling a Cython binary, it
+        # generates Python annotations sufficient to use inspect#signature.
+        sig = _infer_signature_via_inspect(
+            obj=cast(Callable[..., Any], obj),
+            default_ret_type=default_ret_type)
+        if sig is not None:
+            return [sig]
+        # Fall through to parse via doc if inspect.signature() didn't work
+
+    if (name in ('__new__', '__init__') and name not in sigs and class_name and
+            class_name in class_sigs):
+        return [FunctionSig(name=name,
+                            args=infer_arg_sig_from_docstring(class_sigs[class_name]),
+                            ret_type=default_ret_type)]
+
+    docstr = getattr(obj, '__doc__', None)
+    inferred = infer_sig_from_docstring(docstr, name)
+    if inferred:
+        return inferred
+
+    if class_name and name not in sigs:
+        return [FunctionSig(name, args=infer_method_sig(name), ret_type=default_ret_type)]
+    else:
+        return [FunctionSig(name=name,
+                            args=infer_arg_sig_from_docstring(
+                                sigs.get(name, '(*args, **kwargs)')),
+                            ret_type=default_ret_type)]
+
+
+def _infer_signature_via_inspect(obj: Callable[..., Any], default_ret_type: str) -> \
+        Optional[FunctionSig]:
+    """
+    Parses a FunctionSig via annotations found in inspect#signature(). Returns None if
+    inspect.signature() failed to generate a signature.
+    """
+
+    try:
+        signature = inspect.signature(obj)
+    except (ValueError, TypeError):
+        # inspect.signature() failed to generate a signature; this can happen for some methods
+        # depending on the implementation of Python, or if a cython function was not compiled with
+        # binding=True.
+        return None
+    args = []
+
+    def annotation_to_name(annotation: Any) -> Optional[str]:
+        if annotation == inspect.Signature.empty:
+            return None
+        if isinstance(annotation, str):
+            return annotation
+        if inspect.isclass(annotation):
+            return annotation.__name__
+        if hasattr(annotation, '__str__'):
+            return annotation.__str__()
+        # Can't do anything here, so ignore
+        return None
```

Comment:
I wonder what sorts of things Cython might add to the signature. Have you tried that this does something reasonable with all types supported by Cython? It would be good to include all relevant sorts of types in the test case, in case something isn't covered yet.